runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java

```diff
@@ -17,31 +17,173 @@
+  private final ListeningExecutorService executorService;
+  private List<Consumer<Enum>> stateObservers;
+  private List<Consumer<JobMessage>> messageObservers;
+  private JobState.Enum jobState;
+  @Nullable private ListenableFuture<PipelineResult> invocationFuture;
+
+  protected JobInvocation(
+      String id,
+      String retrievalToken,
+      ListeningExecutorService executorService,
+      Pipeline pipeline) {
+    this.id = id;
+    this.retrievalToken = retrievalToken;
+    this.executorService = executorService;
+    this.pipeline = pipeline;
+    this.stateObservers = new ArrayList<>();
+    this.messageObservers = new ArrayList<>();
+    this.invocationFuture = null;
+    this.jobState = JobState.Enum.STOPPED;
+  }
+
+  private PipelineResult runPipeline() throws Exception {
+    return run(pipeline);
+  }
 
   /** Start the job. */
-  void start();
+  public synchronized void start() {
+    LOG.info("Starting job invocation {}", getId());
+    if (getState() != JobState.Enum.STOPPED) {
+      throw new IllegalStateException(String.format("Job %s already running.", getId()));
+    }
+    setState(JobState.Enum.STARTING);
+    invocationFuture = executorService.submit(this::runPipeline);
+    // TODO: Defer transitioning until the pipeline is up and running.
+    setState(JobState.Enum.RUNNING);
+    Futures.addCallback(
+        invocationFuture,
+        new FutureCallback<PipelineResult>() {
+          @Override
+          public void onSuccess(@Nullable PipelineResult pipelineResult) {
+            if (pipelineResult != null) {
+              checkArgument(
+                  pipelineResult.getState() == PipelineResult.State.DONE,
+                  "Success on non-Done state: " + pipelineResult.getState());
+              setState(JobState.Enum.DONE);
+            } else {
+              setState(JobState.Enum.UNSPECIFIED);
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable throwable) {
+            String message = String.format("Error during job invocation %s.", getId());
+            LOG.error(message, throwable);
+            sendMessage(
+                JobMessage.newBuilder()
+                    .setMessageText(getStackTraceAsString(throwable))
+                    .setImportance(JobMessage.MessageImportance.JOB_MESSAGE_DEBUG)
+                    .build());
+            sendMessage(
+                JobMessage.newBuilder()
+                    .setMessageText(getRootCause(throwable).toString())
+                    .setImportance(JobMessage.MessageImportance.JOB_MESSAGE_ERROR)
+                    .build());
+            setState(JobState.Enum.FAILED);
+          }
+        },
+        executorService);
+  }
 
   /** @return Unique identifier for the job invocation. */
-  String getId();
+  public String getId() {
+    return id;
+  }
+
+  /** @return Token to retrieve artifacts staged via the artifact API. */
+  public String getRetrievalToken() {
```

Comment:
Do we need this method? Seems like it is only used within the class to return the JobInfo, so we could get rid off the getter.

Reply:
I'm not sure&mdash;it was part of the old interface, so I kept it. Also, I prefer having a getter instead of letting a subclass a protected field directly.

Reply:
If the field is `protected` then it is fine to be accessed by a subclass. In this case it is even `final`. IMHO getters are overused by many Java developers and getters come with their own sort of problems, e.g. when inheritance is used.

Just something to think about, we can leave it like it is.